src/wayland/wl.c

```diff
@@ -0,0 +1,743 @@
+        } else if (strcmp(interface, zwlr_layer_shell_v1_interface.name) == 0) {
+                ctx.layer_shell = wl_registry_bind(registry, name,
+                        &zwlr_layer_shell_v1_interface, 1);
+        } else if (strcmp(interface, wl_seat_interface.name) == 0) {
+                ctx.seat = wl_registry_bind(registry, name, &wl_seat_interface, 3);
+                wl_seat_add_listener(ctx.seat, &seat_listener, ctx.seat);
+                add_seat_to_idle_handler(ctx.seat);
+        } else if (strcmp(interface, wl_output_interface.name) == 0) {
+                struct wl_output *output =
+                        wl_registry_bind(registry, name, &wl_output_interface, 3);
+                create_output(output, name);
+        } else if (strcmp(interface, zxdg_output_manager_v1_interface.name) == 0 &&
+                        version >= ZXDG_OUTPUT_V1_NAME_SINCE_VERSION) {
+                ctx.xdg_output_manager = wl_registry_bind(registry, name,
+                        &zxdg_output_manager_v1_interface,
+                        ZXDG_OUTPUT_V1_NAME_SINCE_VERSION);
+        } else if (strcmp(interface, org_kde_kwin_idle_interface.name) == 0 &&
+                        version >= ORG_KDE_KWIN_IDLE_TIMEOUT_IDLE_SINCE_VERSION) {
+                ctx.idle_handler = wl_registry_bind(registry, name, &org_kde_kwin_idle_interface, 1);
+        }
+}
+
+static void handle_global_remove(void *data, struct wl_registry *registry,
+                uint32_t name) {
+        struct dunst_output *output, *tmp;
+        wl_list_for_each_safe(output, tmp, &ctx.outputs, link) {
+                if (output->global_name == name) {
+                        destroy_output(output);
+                        break;
+                }
+        }
+}
+
+static const struct wl_registry_listener registry_listener = {
+        .global = handle_global,
+        .global_remove = handle_global_remove,
+};
+
+bool init_wayland() {
+        wl_list_init(&ctx.outputs);
+        //wl_list_init(&ctx.seats);
+
+        ctx.display = wl_display_connect(NULL);
+
+        if (ctx.display == NULL) {
+                LOG_E("failed to create display");
+                return false;
+        }
+
+        ctx.registry = wl_display_get_registry(ctx.display);
+        wl_registry_add_listener(ctx.registry, &registry_listener, NULL);
+        wl_display_roundtrip(ctx.display);
+
+        if (ctx.compositor == NULL) {
+                LOG_E("compositor doesn't support wl_compositor");
+                return false;
+        }
+        if (ctx.shm == NULL) {
+                LOG_E("compositor doesn't support wl_shm");
+                return false;
+        }
+        if (ctx.layer_shell == NULL) {
+                LOG_E("compositor doesn't support zwlr_layer_shell_v1");
+                return false;
+        }
+        if (ctx.seat == NULL) {
+                LOG_W("no seat was found, so dunst cannot see input");
+        }
+        else {
```

Comment:
```suggestion
        } else {
```

Reply:
Else if on the same line as else as well? (see a few lines below)

Reply:
I missed that, good catch. I generally try to follow the [kernel code style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html) with a few deviations (indentation being one of them, as I don't want to break `git blame` with a huge commit updating it).

Reply:
Thanks. that may be helpful in HACKING.md or some contributing doc as well.